### DISCUSSION TITLE
Feature: indi-allsky Docker Container

### DISCUSSION THREAD

**Discussion by spensireli** [aaronwmorris/indi-allsky#570] [Ideas]
I'd be willing to give a crack at Dockerizing this application and writing a compose file for it. Before I did I wanted to check if this wasn't already started somewhere.

**Comment by aaronwmorris** [comment]
I have not done any work related to containerization yet.

**Comment by Elliot-Ford** [comment]
I did some digging into this as this is a feature-set that I'd like to see as well. The biggest initial hurdle that I encountered would be removing and replicating the systemd/dbus calls found in a couple of places (namely the `views.py`).
I've seen some discussion on how one might go about replicating this with a container, but haven't done any experimentation with it yet

**Comment by aaronwmorris** [comment]
I have done some work toward using docker.  You can see what I have in the `docker_testing/` folder.  Nothing ground breaking there.  I also updated the setup.sh script to run non-interactive using settings from environment variables.

@Elliot-Ford is right about the systemd stuff, but with a slightly different spin from my point of view.  Trying to combine all of the services into a single container is not really a [docker] best practice.  It can be done, but it is not the docker way.

indi-allsky could be split into as many as 4 containers:

1. indiserver
2. indi-allsky capture process
3. gunicorn web app
4. Apache/Nginx reverse proxy (this could be optional)

(with shared volumes between 2, 3, and 4)

I suppose nobody would really care about combining all of these into one container if it is just for testing.

I do think that using the setup.sh in the docker build is not going to work.  I think I will just write a custom setup process for the docker container and try to keep the setup.sh and docker builds in sync.

**Comment by aaronwmorris** [comment]
I have merged #573 which moves the contents of `config.json` to the database.  One less file asset that needs to be shared between containers.

**Comment by ennui2342** [comment]
Just to add to this that a containerised version would be interesting with the new syncapi. My allsky has very limited storage in the field, so I backup via FTP to my local nas. However, viewing this archive is a pain, so the new syncapi looks great to have a proper web interface to an archive running on top of my nas. My setup here is a k8s cluster attached to the nas, so spinning up a container would be the ideal way of doing this.

**Comment by aaronwmorris** [comment]
With PR #767 docker support is effectively in place.

There is likely a lot of room for improvement, but with the current docker-compose script, you get a fully functional indi-allsky environment using the indi_simulator_ccd out of the box.  The camera server can be changed, but I have not tested actually connecting a real camera to the dockerized indiserver yet.

**Reply by aaronwmorris** [reply]
On Linux, I have tested an ASI camera and it seemed to work just fine.  The indiserver container runs in privileged mode so it should get access to USB devices.

**Comment by aaronwmorris** [comment]
I originally had issues with building the images on Docker desktop, however I discovered the issue was docker desktop defaults to docker compose v2 where I was doing all of my testing using compose v1.

As of #798, everything is working with compose v2 including on the Mac M1+ platform.
